text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,444 @@
+if_same_then_else = "allow"
+let_and_return = "allow"  # sometimes good to name what you are returning
+multiple_bound_locations = "allow"
+assigning_clones = "allow"
+blocks_in_conditions = "allow"
+```
+
+Also, CI runs `clippy` with `-D warnings -A deprecated`
+- Dependency upgrades should not be blocked on an API becoming deprecated
+- The maintainers want visibility into what is deprecated though
+- Combined with other circumstances, this also blocks [`clippy-sarif`](https://crates.io/crates/clippy-sarif) from opening issues in PRs for deprecated APIs.
+
+Part of this stems from `deprecated` having two purposes:
+- This API is broken and you should severely question any use of it
+- This API will go away with the next major version you should migrate to the new API at some point in time
+
+The second case is very useful for compiler-guided upgrades through breaking changes
+(e.g. [winnow's migration guide](https://github.com/winnow-rs/winnow/blob/v0.6.0/CHANGELOG.md)).
+
+Allowing dependency upgrades in the presence of deprecations is not just something that `clap` is interested in for itself but
+users of clap reported that they want to handle deprecations on their own schedule ([clap#3822](https://github.com/clap-rs/clap/issues/3822)).
+This led to adding a `deprecated` feature to `clap` so deprecations became opt-in
+([clap#3830](https://github.com/clap-rs/clap/pull/3830)).
+However, this can be easy for users to miss and won't catch new uses of the deprecated APIs.
+
+## Example: `cargo`
+
+In 2021 ([cargo#9356](https://github.com/rust-lang/cargo/pull/9356)), the Cargo team switched their `src/lib.rs` to:
+```rust
+#![allow(clippy::all)]
+#![warn(clippy::needless_borrow)]
+#![warn(clippy::redundant_clone)]
+```
+due to false positives and the level of subjectivity of the improvements.
+This approach was first relaxed in 2024 ([cargo#11722](https://github.com/rust-lang/cargo/pull/11722)).
+
+Over time and with the addition of the `[lints]` table, this eventually led to ([cargo#12178](https://github.com/rust-lang/cargo/pull/12178)):
+```toml
+[workspace.lints.clippy]
+all = { level = "allow", priority = -1 }
+dbg_macro = "warn"
+disallowed_methods = "warn"
+print_stderr = "warn"
+print_stdout = "warn"
+self_named_module_files = "warn"
+```
+
+In a recent PR, a bug that clippy would have found without our level overrides (`derive_ord_xor_partial_ord`)
+was only caught in review because one of the reviewers has seen clippy report
+this lint many times in other projects.
+After some discussion, the `Cargo.toml` was updated to ([cargo#14796](https://github.com/rust-lang/cargo/pull/14796)):
+```toml
+[workspace.lints.clippy]
+all = { level = "allow", priority = -2 }
+correctness = { level = "warn", priority = -1 }
+dbg_macro = "warn"
+disallowed_methods = "warn"
+print_stderr = "warn"
+print_stdout = "warn"
+self_named_module_files = "warn"
+```
+
+If more clippy lints were non-blocking, maybe the Cargo team would not have set
```

Comment:
I find this statement to be a bit weird. Most clippy lints _are_ non-blocking. "warn" _is_ non-blocking. Clippy has a small set of blocking lints, which is `correctness`, which is exactly what this config ends up falling to, more or less, except for a `-Dwarnings` world.

Reply:
That statement is made within the context of the start of the motivation section:

> By its name and literal usage, the `warn` level is non-blocking.
> However, most projects treat `warn` as a soft-error.
> It doesn't block for local development but CI blocks it from being merged.
> This is an attempt to balance final correctness with rapid prototyping.
> Requiring "warnings clean" code also avoids warnings fatigue where warnings
> make it hard to see "relevant" compiler output, and make a codebase feel
> lower-quality in a way that does not inspire people or invite people to
> help solve the problem.
> This convention is not new with the Rust community; many C++ projects have take
> this approach before Rust came to be with "warnings clean" being a goal for
> correctness.
> Cargo is looking to further cement this meaning by adding
> [`CARGO_BUILD_WARNINGS=deny`](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#warnings)
> for CIs to set rather than `RUSTFLAGS=-Dwarnings`.

Reply:
Going to post a bigger comment about this. I think there's a subtler thing underlying this claim.

Reply:
## Why do people `-Dwarnings`?

The RFC has an interesting tidbit that inadvertantly reveals one of the subtler things underlying this discussion

> If more clippy lints were non-blocking, maybe the Cargo team would not have set `all` to `allow` and been able to catch this more easily.

The interesting thing is ... **the lints in question are, technically speaking, non-blocking**! They were `warn`! And the choice the `cargo` team made with their codebase here was ultimately to block on the Clippy `correctness` lint group ... **which is `deny` by default**! They sort of arrived at the choices made by clippy _available by default_, the long way around.

...except that's not quite the case in a `-Dwarnings` world, right? In such a world, "warn" counts as blocking. And if you're _used_ to a `-Dwarnings` world, it is a totally consistent thing to claim all of Clippy's `warn` lints as being "blocking", even though that's not a default.

Which makes me step back a bit and think a bit more about why people use `-Dwarnings` with clippy. Because, yes, that is an absolutely supported and valid workflow, and even somewhat encouraged, but it's not the default. If we wanted it to be the default we'd just mark more of our lints as `deny`! But a lot of people CI with `-Dwarnings`, including most codebases I work on.

And it comes back to some principles about warning fatigue. A linter is only useful if it's not noisy. One of the ways a linter can be noisy is a workflow where people do not necessarily run the linter on every commit, so there's plenty of code in the codebase that throws lints, and a contributor adding code will find lints about their changes buried amongst hundreds of lints from the rest of the codebase[^1].

So most people CI with `-Dwarnings` to ensure that the codebase is linter-clean, so that any _new_  warnings are by necessity warnings in new code, and thus relevant.

Which means that, of course, the `-Dwarnings` world is "normal" for most of us, at least in CI.

Given all this, though, do we think that `nit`s will be treated differently?

Stepping through the steps again:

 1. For `nit`s to be useful, your codebase should not be emitting tons of them, and only showing you _relevant_ ones
 2. The easiest way to ensure this is to ensure that the `main` branch is `nit`-free
 3. People are going to want to build with `-Dnits`
 4. We have the same issues as people who build with `-Dwarnings` are having.

... which makes me feel like this is introducing a distinction without a difference, that won't actually solve any of these problems in practice, unless paired with other improvements elsewhere.

I think the underlying problem here is (1): lints are not useful if a codebase is full of them. That could potentially be improved upon for `nit`s, somehow.

For example, we could potentially make it the norm to rarely, if ever use `-Dnits` (or even make it unavailable, though I don't like restricting functionality). Instead, this feature is design in concert with IDEs to provide the right UX such that nits are still useful even if a codebase is not nit-free from the get go. The RFC has a good start of it mentioning the default as being hidden except in IDEs, but I think more holistic UX work could be done there. Either only showing nits on code that has been modified by the current user, or just expecting they be shown with much less weight so that seeing a file full of nits is overall not a big idea.


I also wonder if, as mentioned elsewhere, supporting lint profiles would be a good solution to many of the issues here. Being able to say "by default build with these lint settings, but when building with `--profile morelints` or `--lints morelints` build with these settings instead". Perhaps IDEs are set up to request slightly noisier lints.

As it stands I'm not convinced that `nit` as a feature will in and of itself fix the problems talked about in the motivation here. I think it is a component of many solutions, but not a solution in and of itself.
 



 [^1]: Tooling _can_ help here; there are code review designs where the only lints shown are ones relevant to the actual code being edited. But most people don't use things like that.

Reply:
> I also wonder if, as mentioned elsewhere, supporting lint profiles would be a good solution to many of the issues here. Being able to say "by default build with these lint settings, but when building with --profile morelints or --lints morelints build with these settings instead". Perhaps IDEs are set up to request slightly noisier lints.



thought: this proposal is _basically_ giving the ability to do more fine grained versions of `-Dwarnings`. Because it's _true_ that a codebase wants to run different "sets" of lints at different times:  `-Dwarnings`  during development is terrible, but somewhat normal during dev, and having finer grained control would be amazing.

Reply:
> And it comes back to some principles about warning fatigue. A linter is only useful if it's not noisy. 

I think this is a reductionist take.  The motivation section is not theoretically constructed but coming from my experience with my own code and in coming into another code base and seeing how they work.  I specifically called out these projects.  I specifically called out where the current workflows are failing each project.  "Warning fatigue" was not a pertinent factor in those projects.

I think it would be good to first take a step back and see what basic principles we can both agree on.

Do you agree people have strong enough reason to view lints in the following workflows around lints with users being empowered choose which lints are associated with each workflow?
- This is not worth producing a binary / running tests (I can see putting `derive_ord_xor_partial_ord` in this category)
- Fine for local experimentation but should be cleaned up for production (`dead_code`)
- Help catch little things that may make improvements (`let_and_return`)
- Things I never want to hear about (`mod_module_files` for me because prevent `self_named_module_files`)

Do you know of any other workflows not covered above?

Reply:
I mean, I don't think my point about warning fatigue is a point _against_ this RFC? And it's literally mentioned in the first paragraph of the motivation. I 100% agree with the motivation that it's important to be able to see _relevant_ compiler output! 

This comment is me trying to unravel some of the choices that led to Cargo choosing to arrive at clippy's defaults "the long way around" (i.e., if Cargo had just chosen to run `cargo clippy` in CI with no additional flags). The reasons for Cargo to do that make sense! It's a `-Dwarnings`-in-CI codebase, and it has good reason to use `-Dwarnings`-in-CI, as do most codebases in the Rust ecosystem, and making that choice for its `cargo clippy` run seems natural. Cargo didn't do anything _wrong_ here, but it's definitely an indicator of _something_ wrong that Cargo went through this cycle.


This _is_ my attempt to step back; I'm trying to sketch out the problem space a bit more.


> Do you agree people have strong enough reason to view lints in the following workflows around lints with users being empowered choose which lints are associated with each workflow?


Yes, but I think there's more to it. I think my position is that this four-way distinction (deny, deny-in-CI, informative, hidden) is a good one for many codebases, and I absolutely agree that people have good reasons to have that distinction. I also think that it's not the _only_ such distinction available.

CI can be multifaceted: for example some of people care about denying _rustc_ lints[^1] on all CI'd versions, but will want `-Dwarnings` clippy only on a particular version of the compiler (already somewhat easy to make happen in CI systems). There's lints you only care about during release time if you're not an always-releasable system, which people currently either do `#[allow(lint)] // TODO #1234` or  by perma-allowing the lint and having a release step that's about enabling the lint[^2].

There's also the time axis: a thing I've noticed in large monorepo codebases is that they (or their toolchain) introduce new lints/analyses every now and then, _but they aren't expected to fix old code when they do_. _New_ code however is basically required to fix these. This is a great use case for some of the things you've mentioned in this RFC around warning on things introduced in a PR.

So what I see out there is a distinction of "deny, inform, allow, deny-in-CI, deny-in-release, deny-in-whatever, deny-in-whateverelse, ...". I'm not really convinced the solution here is to just offer _one_ knob that works by ignoring `-Dwarnings` and also having different UX. I see multiple separate concerns here:

 - We just don't have enough control over contextual blocking/non-blocking
 - The UX of warnings is quite rigid and it would be nice to have more differentiation here

## We just don't have enough control over contextual blocking/non-blocking

We have two ways to control what is blocking and non blocking:

 - Setting lint levels
 - `-Dwarnings`

The former is hardcoded (yes, you can pass things in the CLI, but that's not pleasant). The latter is a _single_ toggle that affects every hardcoded "warn" lint wholesale. There's no further control. That kind of sucks. The motivation for this RFC heavily cites cases around this, but I'm not convinced that the solution proposed is sufficient. 

I think focusing on `-Dwarnings` and the reasons people choose it is a good way to come up with a good solution for the problems here. Perhaps you disagree.

## The UX of warnings is quite rigid and it would be nice to have more differentiation here

All warnings are currently alike, and I don't think they really should be in practice. Clippy attempts to make some differentiation with lint grouping, but that doesn't really change the UI/UX: we still just have "error, warning, nothing" to choose from.

I really think a `nit` lint level solves _this_ problem quite well. It would be great to signal to IDEs that they need to do different UI for a lint. We could even improve the rustc TUI around nits by doing a collapsed thing where it doesn't show all warnings, just "also found X nits" or "also found X `let_and_return` nits, and Y `cast_lossless` nits".


-------


Anyway, hope that better illustrates where I'm coming from.




 [^1]: Which are way more careful about new lints across releases
 [^2]: ICU4X doesn't do this for lints, but we do do this for rustfmt'ing doctests. We don't want every PR to be gated on that, it's annoying, but we do have a release checklist step for doing it.

Reply:
> I mean, I don't think my point about warning fatigue is a point against this RFC? And it's literally mentioned in the first paragraph of the motivation. I 100% agree with the motivation that it's important to be able to see relevant compiler output!

I didn't say it was, directly, but I feel like parts of what I'm trying to communicate are being missed because of the focus on backtracking how things came to be and on the implementation.  I'm trying to step back to basics to see if we can agree on those so we have a foundation for the rest of the conversation.

> Yes, but I think there's more to it. I think my position is that this four-way distinction (deny, deny-in-CI, informative, hidden) is a good one for many codebases, and I absolutely agree that people have good reasons to have that distinction. 

Alright, this sounds like a good place to start!

> some of people care about denying rustc lints[1](https://github.com/rust-lang/rfcs/pull/3730#user-content-fn-1-1e933df256da21d625b0adb31475f1fa) on all CI'd versions, but will want -Dwarnings clippy only on a particular version of the compiler (already somewhat easy to make happen in CI systems).

I assume users deny on a specific version for clippy so master stays green.  if they are already handling that, why bother denying rustc lints in all releases rather than just doing it in the one clippy job and avoid redundant runs?  If its denying while building for tests, then that means they can't get any test feedback until all style warnings are addressed which adds an unnecessary sequencing that slows down feedback to the user.

>  There's lints you only care about during release time if you're not an always-releasable system, which people currently either do #[allow(lint)] // TODO #1234 or by perma-allowing the lint and having a release step that's about enabling the lint

Good point for us to consider this workflow.  For myself, I think we shouldn't go out of our way to support this workflow and covered that in d5814384.

> So what I see out there is a distinction of "deny, inform, allow, deny-in-CI, deny-in-release, deny-in-whatever, deny-in-whateverelse, ...". I'm not really convinced the solution here is to just offer one knob that works by ignoring -Dwarnings and also having different UX. I see multiple separate concerns here:

Do we need to support every solution?

In Cargo, our [design principles](https://doc.crates.io/contrib/design.html) push for a canonical workflow, rather than supporting every workflow.  I know Rustc and clippy are different and need to focus more on the mechanisms of doing things rather than end-user policy like this.

I feel like we also have other biases that push me to this solution
- We now have the `[lints]` table and people are already using levels in that to get most of the semantics and to handle other semantics through a wholly new solution would be disruptive and is a lot larger of a new design to work through
- Cargo has an unstable replacement for `-Dwarnings` that, while a mechanism, smoothing out the path for warnings being soft errors

It could be said that if we get the right mechanisms, we could allow us to invent new, unexpected workflows, in the future.  However, over-generalization also has costs.

Reply:
> I feel like parts of what I'm trying to communicate are being missed because of the focus on backtracking how things came to be and on the implementation.

I went through that exercise because I noticed that Cargo had gone through a bunch of steps to arrive at Clippy's original defaults, and I didn't want us to add a new feature that leads to the same dynamic just at a different level. I'm still not convinced that the `nit` lint level (with no other changes) won't eventually end up in the same cycle of "people want to `-Dnit`" -> "people are frustrated when things are too noisy".


> Good point for us to consider this workflow. For myself, I think we shouldn't go out of our way to support this workflow and covered that in https://github.com/rust-lang/rfcs/commit/d5814384b304a9a2baa2f5f4a91546fdab2025c7.

To be clear, it was an example, I think there are a lot of other things in this space. Like, for example "lints for people learning". This is a topic that has come up countless times and the current lint level architecture is just insufficient for it because we just don't support wholesale modalities beyond `-Dwarnings`. 


>if they are already handling that, why bother denying rustc lints in all releases rather than just doing it in the one clippy job and avoid redundant runs?  

As I mentioned in a different comment, clippy lint levels have a lower severity than Rust ones. I think it's a legitimate use case to want to be warnings-green for _rust_ on all versions but for _clippy_ only on one particular version. _Especially_ due to new Rust lints often signaling future incompatibility / edition stuff.

ICU4X has missed out on future incompatibility stuff until it was much later in the cycle, in the past.

> Do we need to support every solution?


No, but I think it's somewhat suboptimal to tack things onto the single modality switch that is `-Dwarnings` hoping we're supporting most use cases. Like yeah, Cargo is adding a Newer, Better, `-Dwarnings` in the form of `CARGO_BUILD_WARNINGS=deny`, but then someone's going to ask for "let me `-Dnits`", so we'll need `CARGO_NITS_WARNINGS=deny`. The solution seems a bit piecemeal rather than holistic.


> We now have the [lints] table and people are already using levels in that to get most of the semantics and to handle other semantics through a wholly new solution would be disruptive and is a lot larger of a new design to work through

I guess I'm coming from the same angle? I think `[lints]` can be extended really well to have this kind of configurability! You said it yourself in the RFC: Clippy `pedantic` serves the purpose of the type of lint talked of in this RFC, kind of, except you can't manually add things to the group. I actually do think that "giving users the ability to group lints into profiles" and allowing profiles to be set globally is a really promising route.

I think what I am talking about is less of a departure from the current workflow than a new "nit" lint level. It might be more work to implement, but I think it'll support more use cases in the long run.

(As said before, I still want a nit lint level, but designed as a solution to a different set of problems)

At the very least I do think it's worth exploring the space of modalities for lints and solutions for that space, and then perhaps concluding that a "nit" lint level is in fact the best solution to pick, trading off use cases supported, implementation complexity, and migration complexity.

I'm happy to write up an RFC for this if you don't want to.

Just as a data point: I have a hard time thinking Clippy will move its `pedantic` group over to `nit` (though maaaybe we'd split it? Unsure). So people wishing for this kind of configurability with clippy are going to have to add to `[lints]` anyway, in some form. I'm thinking something more structured would be nice.

Reply:
Are we still assuming that `nit` only runs against new code? If that's the case, it's essentially impossible for it to be noisy.

Reply:
Yes, I am, but I don't think it's impossible for it to be noisy. "new code" is firstly a bit of a nebulous thing, I've found that systems which do this typically do it for all files touched, not narrowed to all modified lines. It's unclear to me if that choice is because lints often fire _near_ changed code, or because this serves as a nudge to fix other issues in the file.

Which is still not prone to noise except ... new lints are a thing. And new lints _do_ get noisy since they hit you like a flood (remember `check-cfg`? And of course this is a regular experience of people who run clippy with `-Dclippy::pedantic`, but that's somewhat expected) even when you filter down to only changed files. Perhaps not if you filter down to changed lines of codes? Unclear.

But thanks for the reminder: there have been a lot of ideas of this type in the past and it's easy to get muddled. This proposal does not propose any actual surfacing of `nit`s in first-party Rust tooling outside of json output. 

(I actually _would_ like a `nit` lint level that has better UX in `rustc`, potentially collapsing down to a single informative error listing the number of nits of each kind)

Reply:
I'm assuming that even a new nit would only fire against new code, not all code, since it's still just a nit. So the amount of code that could possibly have nits firing on it is still capped to "code touched by the current PR or git diff". And so, assuming people understand their own PRs (maybe this shouldn't be assumed), the total nit level should stay within range of the programmer writing the PR.

Reply:
Just another example of lint modalities: "warn for main code but not tests". ICU4X does this and because it does this we cannot use the `[lints]` table right now.

```
    #![cfg_attr(
        not(test),
        deny(
            clippy::indexing_slicing,
            clippy::unwrap_used,
            clippy::expect_used,
            clippy::panic,
            clippy::exhaustive_structs,
            clippy::exhaustive_enums,
            missing_debug_implementations,
        )
    )]
```


(Ignore that this is `deny`: it could be `warn` as well and would not change our workflow/CI in any way, that's just the choice that was made when this boilerplate was first written. We `-Dwarnings --all-targets --all-features` in Clippy CI.)


For the clippy restriction lints[^1] it doesn't really make sense to have them enabled on tests, and the same goes for `missing_debug_implementations`, so that's a type of useful modality to have available. Ideally one that gets automatically handled (a "tests" lint profile that's automatically enabled when building test-only code)  but not necessarily.

 [^1]: A restriction lint is effectively "this is not considered bad, however codebases may have reasons to wish to disallow it". ICU4X is no-panics so it tries to flag as many of those as possible, and its API evolution rules also require an exhaustive default. Neither of these are things that matter for most codebases, and they also do not matter for ICU4X test code (we _want_ test code to be able to panic, of course).

Reply:
> "code touched by the current PR or git diff

... or at least _files_ touched by the current PR.

It should stay within range for fixing, but if it's annoying enough people may still disable it and want to do a whole-program fix sweep before it gets enabled.

I do agree that "nits aren't shown in most default workflows" _mostly_ solves the noisiness problem.

Reply:
I mean *personally* I don't want the CI to do stuff different from my own computer, so I'd either turn on nits on my own machine all the time, or I'd turn them all off forever. I'm not exactly the target audience of this RFC.

Reply:
@Lokathor Okay, but this is exactly my worry here :smile: . I'm in a similar boat: I don't use an IDE, and I don't want my CI to be _that_ different from the local computer[^1], I just want "what is blocking" to be different. I don't think this RFC solves the problems it is motivated by for people like me.


I'm the target audience for the motivation. I'm not the target audience for the solution. That is disappointing, and I do not think that means we should not participate in this RFC.

 [^1]: Too many experiences with complex codebases and being unable to figure out why something is broken in CI and not locally. Almost all of ICU4X's CI jobs are specified as `cargo make ci-job-foo` where `ci-job-foo` has the actual subtasks, so if a CI task fails and you can't figure out why, worst case you can just run the `cargo-make` task locally. The "make clippy show up on PRs" task, and deploy tasks, are the only ones that do something different, and I have had to debug the clippy one before.

Reply:
> I actually do think that "giving users the ability to group lints into profiles" and allowing profiles to be set globally is a really promising route.

This and giving tools like Clippy the ability to pre-define profiles would be a great feature. This would also cover requests like https://github.com/rust-lang/rust-clippy/issues/5537. 

> Just as a data point: I have a hard time thinking Clippy will move its `pedantic` group over to `nit` (though maaaybe we'd split it? Unsure).

I also don't think we would move all of `pedantic` to the `nit` lint level. I can see us introduce a new lint group in Clippy , where we move some hand-picked lints from `pedantic` and `style` to (maybe also from some other groups). But that's a decision we can make, if the `nit` level gets introduced.

> I'm in a similar boat: I don't use an IDE, and I don't want my CI to be _that_ different from the local computer

+1. I'm probably one of the persons who would set (most) nit lints to warnings and then let `-Dwarnings` on CI deal with that. And the `nit` lints that I don't want, I would `allow` in the `Cargo.toml`. I think there would be a narrow selection of `nit` lints that I would actually keep as `nit`. So I think introducing a `nit` lint level would be useful in general.

Reply:
I think there are two important questions
- If this doesn't work for me, is there a reasonable solution that includes me?
- If this still doesn't work for me, is there still enough value in the RFC to benefit others?

Also, another interesting characteristic of this solution that I forgot to call out earlier in this thread is that the displaying of these lints is being handled in a layer above rustc.  This means a user or tool can show or hide these without recompilation as Cargo logs messages from previous runs.  Currently, anything else would thrash the cache though hopefully that will change soon (rust-lang/cargo#14830).

Reply:
> As I mentioned in a different comment, clippy lint levels have a lower severity than Rust ones. I think it's a legitimate use case to want to be warnings-green for rust on all versions but for clippy only on one particular version. Especially due to new Rust lints often signaling future incompatibility / edition stuff.

While I'm ok with clippy being more opinionated and changing things more frequently than the compiler, I suspect the notion that users should treat lint clippy lint levels differently than rust lint levels is not recognized by all clippy users.  I've seen articles in the past recommending to only check `cargo clippy` for lints, rather than `cargo check && cargo clippy` because the second includes everything from the first.  For myself, I think its a UX problem for the Rust project to be inconsistent on this and require users to track the relative meaning of each linters lint levels.  We should have one set of semantics for lint levels across all tools, even if the tools vary in other criteria (opinionated, rate of change).

Reply:
But `cargo clippy` *does* cover everything that a check and then a clippy covers? And each lint says if it's coming from clippy.

Reply:
Yes but what I'm understanding from Manish is that the two tools are disagreeing on what the lint levels mean.  From that perspective, `cargo clippy` isn't a strict superset because you should respond to warnings between them differently.  In my opinion, that puts an unnecessary burden on users that they are unlikely aware of and is counter productive.

Reply:
Oh, yeah, Manish is correct. You're "supposed" to just turn off basically any clippy lint you disagree with, because most of them are pure style over substance.

Clippy, like rustfmt, doesn't always actually have the right defaults, and should be readily configured to fit the project's needs.

Reply:
> it's basically a non-starter to do anything else. Clippy _cannot exist_ with Rust's lint levels. Or it would, but with 99% of its lints being `allow`. Rust has a pretty high bar for lint levels, for quite understandable reasons.
> 
> The idea is that because clippy is something you _choose_ to run later and doesn't block compilation, it can do something different. I think this leads to some confusion but overall it works pretty well.

I think this is coupling a several concepts that I think we can separate
- Opinionated-ness of lints
- Rate of change of lints
- The semantic meaning of lint groups (what is a "soft error")

The last one I think users should be able to treat in a consistent manner.  It should be fine for users to run `RUSTFLAGS=-Dwarnings cargo clippy`.  Because of the rate of change, they just have a higher likelihood of needing to pin the version of clippy used for doing this.  I used to use MSRV for this but wanted the benefit of the latest clippy lints and switched to "latest stable" with Renovatebot updating the version every time a new stable comes up.

> And yeah, I don't think this would be something either ICU4X or Servo-back-when-I-worked-on-it would find useful

How come?

> I do think lint profiles would be a reasonable solution that includes me, I also think there are also other ways to fix the set of problems in the space of "lint modalities" that don't involve profiles. I have a hard time figuring out how to make this RFC fit my needs, but I perhaps haven't thought enough about it.

I have a hard time envisioning this without seeing problems that make me turn away from it.  Maybe you are seeing something different.  Willing to sketch it out?  Also more likely not to be a strawman solution then.

Reply:
> The last one I think users should be able to treat in a consistent manner. It should be fine for users to run `RUSTFLAGS=-Dwarnings cargo clippy`. 

Sure, I think that's the case already. As mentioned before this is standard procedure in a lot of CI (almost always pinned). I think `-Dwarnings cargo clippy` is a lot more work to deal with than plain `cargo clippy` but that's a choice you make when you choose to use clippy. Both choices are equally valid.

I don't think clippy is that inconsistent about the semantic meaning of lint levels. I just think it's much much more different about severity and opinionatedness.

> Also more likely not to be a strawman solution then.

FWIW this isn't something I just came up with, it's been discussed before in various contexts. It seems to come up a lot when people talk about "teaching lints" (I remember I think @llogiq wanting to do teaching lints at some point? could be misremebering) though that needs more design than just "add lint profiles".

I'll sketch out a hackmd.

Reply:
@epage hackmd written: https://hackmd.io/@Manishearth/BJi8K8AGJg

cc @alice-i-cecile you mentioned bevy's use cases and how using Clippy's categories is a non starter: would this proposal work? It essentially allows users to specify their own (cross-workspace if desired) groupings of lints in Cargo.toml and toggle them wholesale.

Reply:
Yes, this would meet our needs, both for the custom linter and for the project as a whole. It would be sweet to be able to share and export them for users via crates.io somehow, but that's easy to do in follow-up.

Reply:
@ijackson thank you, that is useful feedback! From my own use cases Option 2 isn't necessary (I think?) but it's weird to design two axes of a feature (being able to have multiple switchable profiles, and being able to control test-only stuff) and not have them be mixable in a useful way. So it's good to know if people do care about the mixing of the two.

Reply:
> Also, there should be a way to declare some apparently-"production" code to be treated as a test, and vice versa

I suspect the way to get such behavior would be to be able to define custom lint _groups_ that can be toggled in code. That could be built on top of profiles but needs a bunch of rustc side work. I'll list it in a future work section.